File: joint_angles.py
"""
Joint Angle Calculator – computes flexion angles for all 15 finger joints.

Uses the law of cosines on the two bone vectors meeting at each joint.
Result: 15 angles (3 per finger) in degrees (0°–180°).
"""

import math

# Each definition: (name, proximal_index, center_index, distal_index)
# Uses MediaPipe hand landmark indices (0–20).
JOINT_ANGLE_DEFS: list[tuple[str, int, int, int]] = [
    # Thumb (wrist→CMC→MP→IP→Tip = indices 0,1,2,3,4)
    ("thumb_cmc",  0, 1, 2),
    ("thumb_mp",   1, 2, 3),
    ("thumb_ip",   2, 3, 4),
    # Index (wrist→MCP→PIP→DIP→Tip = 0,5,6,7,8)
    ("index_mcp",  0, 5, 6),
    ("index_pip",  5, 6, 7),
    ("index_dip",  6, 7, 8),
    # Middle (0,9,10,11,12)
    ("middle_mcp", 0,  9, 10),
    ("middle_pip", 9, 10, 11),
    ("middle_dip", 10, 11, 12),
    # Ring (0,13,14,15,16)
    ("ring_mcp",   0, 13, 14),
    ("ring_pip",  13, 14, 15),
    ("ring_dip",  14, 15, 16),
    # Little (0,17,18,19,20)
    ("little_mcp",  0, 17, 18),
    ("little_pip", 17, 18, 19),
    ("little_dip", 18, 19, 20),
]


def compute_joint_angles(landmarks) -> dict[str, float]:
    """Compute all 15 flexion angles from MediaPipe landmarks.

    Args:
        landmarks: MediaPipe NormalizedLandmarkList (list-like, 21 items).

    Returns:
        Dict mapping angle name (e.g. "index_pip") to degrees (0–180).
    """
    result: dict[str, float] = {}

    for name, prox_i, cent_i, dist_i in JOINT_ANGLE_DEFS:
        pA = landmarks[prox_i]
        pB = landmarks[cent_i]
        pC = landmarks[dist_i]

        # Vectors: BA = proximal - center, BC = distal - center
        ba_x = pA.x - pB.x
        ba_y = pA.y - pB.y
        bc_x = pC.x - pB.x
        bc_y = pC.y - pB.y

        dot = ba_x * bc_x + ba_y * bc_y
        mag_ba = math.hypot(ba_x, ba_y)
        mag_bc = math.hypot(bc_x, bc_y)

        if mag_ba < 1e-9 or mag_bc < 1e-9:
            continue

        cos_angle = max(-1.0, min(1.0, dot / (mag_ba * mag_bc)))
        degrees = math.degrees(math.acos(cos_angle))
        result[name] = round(degrees, 1)

    return result

File: ios/HandGestureRecognizer/JointAngles.swift
import Foundation
import Vision

// MARK: - Joint angle definitions

/// Each finger has 3 measurable flexion angles (MCP, PIP, DIP).
/// The thumb uses CMC, MP, IP instead — same idea, different bone names.
///
/// The angle at a joint is computed from the three landmarks that bracket it:
///   proximal → joint → distal
/// using the law of cosines on vectors (proximal→joint) and (distal→joint).
struct JointAngleDefinition {
    let name: String
    let proximal: VNHumanHandPoseObservation.JointName
    let center:   VNHumanHandPoseObservation.JointName
    let distal:   VNHumanHandPoseObservation.JointName
}

/// All 15 joint angle definitions (3 per finger).
let allJointAngleDefinitions: [JointAngleDefinition] = [
    // Thumb (CMC → MP → IP → Tip)
    JointAngleDefinition(name: "thumb_cmc",  proximal: .wrist,    center: .thumbCMC, distal: .thumbMP),
    JointAngleDefinition(name: "thumb_mp",   proximal: .thumbCMC, center: .thumbMP,  distal: .thumbIP),
    JointAngleDefinition(name: "thumb_ip",   proximal: .thumbMP,  center: .thumbIP,  distal: .thumbTip),
    // Index (MCP → PIP → DIP → Tip)
    JointAngleDefinition(name: "index_mcp",  proximal: .wrist,    center: .indexMCP, distal: .indexPIP),
    JointAngleDefinition(name: "index_pip",  proximal: .indexMCP, center: .indexPIP, distal: .indexDIP),
    JointAngleDefinition(name: "index_dip",  proximal: .indexPIP, center: .indexDIP, distal: .indexTip),
    // Middle
    JointAngleDefinition(name: "middle_mcp", proximal: .wrist,      center: .middleMCP, distal: .middlePIP),
    JointAngleDefinition(name: "middle_pip", proximal: .middleMCP,  center: .middlePIP, distal: .middleDIP),
    JointAngleDefinition(name: "middle_dip", proximal: .middlePIP,  center: .middleDIP, distal: .middleTip),
    // Ring
    JointAngleDefinition(name: "ring_mcp",   proximal: .wrist,    center: .ringMCP, distal: .ringPIP),
    JointAngleDefinition(name: "ring_pip",   proximal: .ringMCP,  center: .ringPIP, distal: .ringDIP),
    JointAngleDefinition(name: "ring_dip",   proximal: .ringPIP,  center: .ringDIP, distal: .ringTip),
    // Little
    JointAngleDefinition(name: "little_mcp", proximal: .wrist,      center: .littleMCP, distal: .littlePIP),
    JointAngleDefinition(name: "little_pip", proximal: .littleMCP,  center: .littlePIP, distal: .littleDIP),
    JointAngleDefinition(name: "little_dip", proximal: .littlePIP,  center: .littleDIP, distal: .littleTip),
]

// MARK: - Angle computation

/// Compute all 15 flexion angles (in degrees, 0°–180°) from a landmarks dict.
/// Returns a dictionary keyed by angle name (e.g. "index_pip" → 142.3).
/// Joints with missing landmarks are omitted.
func computeJointAngles(
    from landmarks: [VNHumanHandPoseObservation.JointName: CGPoint]
) -> [String: Double] {
    var result: [String: Double] = [:]

    for def in allJointAngleDefinitions {
        guard let pA = landmarks[def.proximal],
              let pB = landmarks[def.center],
              let pC = landmarks[def.distal] else { continue }

        // Vectors: BA = proximal - center, BC = distal - center
        let baX = Double(pA.x - pB.x)
        let baY = Double(pA.y - pB.y)
        let bcX = Double(pC.x - pB.x)
        let bcY = Double(pC.y - pB.y)

        let dot = baX * bcX + baY * bcY
        let magBA = (baX * baX + baY * baY).squareRoot()
        let magBC = (bcX * bcX + bcY * bcY).squareRoot()

        guard magBA > 1e-9, magBC > 1e-9 else { continue }

        // Clamp to [-1, 1] to handle floating-point edge cases
        let cosAngle = min(max(dot / (magBA * magBC), -1.0), 1.0)
        let degrees = acos(cosAngle) * 180.0 / .pi

        result[def.name] = degrees
    }

    return result
}
